swaps/swaps_service.go

```diff
@@ -125,254 +183,1057 @@ func (svc *swapsService) EnableAutoSwaps(ctx context.Context, lnClient lnclient.
+		return err
+	}
+
+	logger.Logger.WithFields(logrus.Fields{
+		"swapId":    swapId,
+		"claimTxId": claimTxId,
+	}).Info("Claim transaction broadcasted for refund")
+
+	err = svc.db.Model(&swap).Updates(&db.Swap{
+		ClaimTxId:     claimTxId,
+		ReceiveAmount: refundAmount,
+		State:         constants.SWAP_STATE_REFUNDED,
+	}).Error
+	if err != nil {
+		logger.Logger.WithFields(logrus.Fields{
+			"swapId":    swapId,
+			"claimTxId": claimTxId,
+		}).WithError(err).Error("Failed to save claim txid to swap")
+		return err
+	}
+
+	return nil
+}
+
+func (svc *swapsService) GetSwap(swapId string) (*Swap, error) {
+	var swap db.Swap
+	err := svc.db.Limit(1).Find(&swap, &db.Swap{
+		SwapId: swapId,
+	}).Error
+
+	if err != nil {
+		logger.Logger.WithError(err).Error("Failed to get swap")
+		return nil, err
+	}
+
+	return &swap, nil
+}
+
+func (svc *swapsService) ListSwaps() ([]Swap, error) {
+	var swaps []db.Swap
+	err := svc.db.Find(&swaps).Error
+
+	if err != nil {
+		logger.Logger.WithError(err).Error("Failed to list swaps")
+		return nil, err
+	}
+
+	return swaps, nil
+}
+
+func (svc *swapsService) subscribePendingSwaps() {
+	var swaps []db.Swap
+	if err := svc.db.Where("state = ?", constants.SWAP_STATE_PENDING).Find(&swaps).Error; err != nil {
+		logger.Logger.WithError(err).Error("failed to load pending swaps")
+		return
+	}
+	if len(swaps) == 0 {
+		return
+	}
+
+	logger.Logger.WithField("count", len(swaps)).Info("Resuming pending swaps...")
+
+	for _, swap := range swaps {
+		var boltzWs *boltz.Websocket
+		for {
+			boltzWs = svc.boltzApi.NewWebsocket()
+			if err := boltzWs.Connect(); err != nil {
+				logger.Logger.WithError(err).Error("Could not connect to Boltz websocket")
+				time.Sleep(5 * time.Second)
+				continue
 			}
-			parsedStatus := boltz.ParseEvent(update.Status)
+			break
+		}
+		if err := boltzWs.Subscribe([]string{swap.SwapId}); err != nil {
+			logger.Logger.WithError(err).WithField("swapId", swap.SwapId).Error("Failed to subscribe to boltz updates")
+			continue
+		}
+		switch swap.Type {
+		case constants.SWAP_TYPE_IN:
+			go svc.startSwapInListener(&swap, boltzWs)
+		case constants.SWAP_TYPE_OUT:
+			go svc.startSwapOutListener(&swap, boltzWs)
+		}
+	}
+}
 
-			switch parsedStatus {
-			case boltz.SwapCreated:
-				logger.Logger.WithFields(logrus.Fields{
-					"swap":   swap,
-					"update": update,
-				}).Info("Paying the swap invoice")
-				err := lnClient.SendPaymentProbes(ctx, swap.Invoice)
-				if err != nil {
-					logger.Logger.WithField("swapId", swap.Id).Info("Couldn't probe invoice payment, terminating swap")
-					return err
+func (svc *swapsService) startSwapInListener(swap *db.Swap, boltzWs *boltz.Websocket) {
+	network, err := boltz.ParseChain(svc.cfg.GetNetwork())
+	if err != nil {
+		logger.Logger.WithError(err).WithFields(logrus.Fields{
+			"swapId": swap.SwapId,
+		}).Error("Failed to parse network")
+		return
```

Comment:
In both the listeners we just return if there's an error. Shouldn't we also set the swap state? should we use defer like we do elsewhere?